x/ccv/parent/keeper/proposal.go

```diff
@@ -42,10 +46,49 @@ func (k Keeper) CreateChildClient(ctx sdk.Context, chainID string, initialHeight
 	if err != nil {
 		return err
 	}
+
 	k.SetChildClient(ctx, chainID, clientID)
+	childGen, err := k.MakeChildGenesis(ctx)
+	if err != nil {
+		return err
+	}
+
+	k.SetChildGenesis(ctx, chainID, childGen)
 	return nil
 }
 
+func (k Keeper) MakeChildGenesis(ctx sdk.Context) (gen childtypes.GenesisState, err error) {
+	unbondingTime := k.stakingKeeper.UnbondingTime(ctx)
+	height := clienttypes.GetSelfHeight(ctx)
+
+	clientState := k.GetTemplateClient(ctx)
+	clientState.ChainId = ctx.ChainID()
+	clientState.LatestHeight = height //(+-1???)
+	clientState.TrustingPeriod = unbondingTime / 2
+	clientState.UnbondingPeriod = unbondingTime
+
+	consState, ok := k.clientKeeper.GetSelfConsensusState(ctx, height)
+	if !ok {
+		return gen, sdkerrors.Wrapf(clienttypes.ErrConsensusStateNotFound, "self consensus state not found for height: %s", height)
+	}
+
+	gen.Params.Enabled = true
+	gen.NewChain = true
+	gen.ParentClientState = clientState
+	gen.ParentConsensusState = consState.(*ibctmtypes.ConsensusState)
+
+	vals := k.stakingKeeper.GetBondedValidatorsByPower(ctx)
```

Comment:
You should use the `LastValidatorPower` index instead, see https://github.com/cosmos/cosmos-sdk/blob/master/x/staking/genesis.go#L193. This contains the last validator set provided to Tendermint. It was provided at the end of the previous block, which means it's in the header of the current block as `NextValidatorsHash`. 

The reason is that by the end of the previous block, the block was already decided, thus the header cannot be changed. As a result, we have these delay of one block: The validator updates provided to Tendermint at height `h`, will be visible in the header as `NextValidatorsHash` at height `h+1` and the updated validator set will validate at hheight `h+2`.

Reply:
Hey @mpoke I have used the method you suggested but there is one edge case. `LastPowers` includes the operator address, but not the consensus address. We need the consensus address to send to the consumer chain, so I use `stakingKeeper.GetValidator` to retrieve it. I am wondering whether it is possible under some scenarios (validator leaving the set the same block?), for this validator not to be found.

Also, we use `stakingKeeper.GetValidatorUpdates` to send the updates during CCV operation- why not use it here? I'm guessing those updates are from the wrong block, right?

Reply:
That's actually very interesting. In the function `CreateChildClient()` (see above), the proposer creates a consensus state for the consumer chain (i.e., a root of trust):
```go
consensusState := ibctmtypes.NewConsensusState(ctx.BlockTime(), commitmenttypes.NewMerkleRoot([]byte(ibctmtypes.SentinelRoot)), ctx.BlockHeader().NextValidatorsHash)
```
In my understanding,  `ctx.BlockHeader().NextValidatorsHash` MUST be the initial validator set on the consumer chain. Thus, if the current height is `h`, then `NextValidatorsHash` is the hash of the validators validating on the provider chain at height `h+1`. 

Now, in `MakeChildGenesis()`, the provider creates the consumer genesis state, which MUST contain the initial validator set, i.e., the one validating on the provider at `h+1`. This is the set updated by the staking module at height `h-1`. Thus, it is part of the  `LastValidatorPowers` as long as `MakeChildGenesis()` is **_called before_** the staking module `EndBlock()` for the current height `h`. 

If I'm not mistaken, `CreateChildChainProposal` (which ends up calling `MakeChildGenesis`) is called in the `EndBlock()` of the gov module, i.e., https://github.com/cosmos/cosmos-sdk/blob/056db1295768cbf4675b8025b66d83bbaadd923b/x/gov/abci.go#L73. I'm not sure though which `EndBlock` has higher priority.